### PR TITLE
Project Management System MVP

### PR DESCRIPTION
**Pull request by dab10**
1. Task:
https://github.com/rolling-scopes-school/tasks/blob/master/tasks/react/project-management-system-EN.md
2.
Снимок экрана (139)
3. Deploy: https://kanban-rss.netlify.app/
4. Video: https://youtu.be/jDgqK-EhD7c
5. Done 04.12.22 (deadline 06.12.22)
6. Score: 60/60

Welcome route - max 7 points
  - [x] The welcome page should contain general information about the developer, project, and course. 1 point
  - [x] In the upper right corner there are 2 buttons: Sign In and Sign Up. 1 point
  - [x] If login token is valid and unexpired, change buttons Sign In and Sign Up on "Go to Main Page" button. 2 points
  - [x] When the token expires - the user should be redirected to the "Welcome page" automatically. 2 points
  - [x] Pressing the Sign In / Sign up button redirects a user to the route with the Sign In / Sign up form. 1 point
Sign In / Sign Up - max 8 points
  - [x] Buttons for Sign In / Sign Up / Sign Out are everywhere where they should be 2 points
  - [x] Form fields should be implemented according to the backend API. Validation should be implemented. 4 points
  - [x] Upon successful login, the user should be redirected to "Main route" 1 point
  - [x] If user already logged in and he try to reach this routes - he should be redirected to Main route. 1 point
Main route - max 8 points
  - [x] Board creation functionality 2 points
  - [x] Displays all created boards as a list/grid 1 point
  - [x] Each board in the list is displayed with a small preview of available information (title, description, etc). By clicking an element the user navigates to the board item (Board route). There's also a button for board deletion. 1 point
  - [x] When trying to delete the board, we should receive a confirmation modal. The confirmation modal must be a generic component (one for the entire application). 1 points
  - [x] The user profile editing functionality is implemented. 3 points
Board route - max 26 points
  - [x] Button for column creation is displayed 1 point
  - [x] If a board contains at least one column - a button for task creation is displayed/become enabled as well 1 points
  - [x] A modal windows with forms is displayed for column and task creations 3 points
  - [x] A vertical scrollbar is displayed in the column when overflowing with the number of column tasks 2 points
  - [x] The page itself on the current route doesn't have a vertical scrollbar 1 points
  - [x] With the help of drag-n-drop, we can swap columns. 3 points
  - [x] With the help of drag-n-drop, we can change the order of tasks within a column. 3 points
  - [x] With the help of drag-n-drop, we can change the task belonging to the column. 5 points
  - [x] The functionality of viewing and editing of the task has been implemented. 3 points
  - [x] The task must have a delete task button. On click: confirmation modal -> delete. 1 points
  - [x] At the top of the column should be Title. When you click on it, it should become an input, with Submit and Cancel buttons near it. After entering text in the input and clicking Submit - the Title of the column should change. 2 points
  - [x] The column should have a delete button. By clicking -> confirmation modal -> when approving -> deleting. 1 points
General requirements - max 11 points
  - [x] Backend error handling - (Not found, unhandled rejection, etc) should be performed in a user-friendly way (toast, pop-up or anything else you implement). 4 points
  - [x] Localization 1 point
  - [x] Backend is deployed and publicly available 2 points
  - [x] Sticky header 2 points
  - [x] Extra scope same complexity as Global search 2 points
Penalties
  - [ ] React default favicon - 1 points
  - [ ] The presence of errors and warnings in the console - 2 points for each
  - [ ] The presence in the console of the results of the console.log execution - - 2 points for each
  - [ ] @ts-ignore or any usage (search through github repo) - - 1 point for each
  - [ ] Making commits after the deadline - - 20 points

### CODE REVIEW

**Review by ardashevdanil — APPROVED** [review]
В целом ок, есть небольшие замечания

**Review by elem15 — APPROVED** [review]
Ok

File: src/app/reducers/authSlice.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+  message: string;
+  messageForAuth: string;
+};
+
+const initialState: IAuthState = {
+  name: '',
+  login: localStorage.getItem('login') || '',
+  password: '',
+  userId: '',
+  token: localStorage.getItem('token') || '',
```
Comment (ardashevdanil) [inline src/app/reducers/authSlice.ts:36]: Лучше из `localStorage` не дёргать сразу в `initialState`, на случай если стейт надо будет сбросить. Лучше из `localStorage` на этапе маунта доставать данные и класть в стор

File: src/app/reducers/boardSlice.ts [hunk 2]
```diff
@@ -0,0 +1,417 @@
...
+  addTasksToColumn,
+  removeColumnsState,
+  addColumns,
+  addTasks,
+  updateTaskInColumn,
+  deleteTaskInColumn,
+  updateColumn,
+} = boardSlice.actions;
+
+export default boardSlice.reducer;
```
Comment (ardashevdanil) [inline src/app/reducers/boardSlice.ts:417]: Большой `slice`, можно было бы разбить на несколько

File: src/components/Burger/Burger.scss [hunk 3]
```diff
@@ -0,0 +1,76 @@
+.burger {
+  z-index: 1000;
```
Comment (ardashevdanil) [inline src/components/Burger/Burger.scss:2]: Обычно берут кратно 10 (10, 20, 30)

File: src/components/EditColumnTitle/EditColumnTitle.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+  const buttons = (
+    <Form.Item>
+      <Button htmlType="submit" icon={<CheckOutlined />} className="submit-button"></Button>
+      <Button onClick={handleOnClickReject} icon={<CloseOutlined />} danger></Button>
+    </Form.Item>
+  );
+
+  const handleOnFocus = () => setOpen(true);
+  const handleOnClick = () => setOpen(true);
```
Comment (ardashevdanil) [inline src/components/EditColumnTitle/EditColumnTitle.tsx:52]: Две одинаковые функции

File: src/components/Layout/Layout.tsx [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import { Outlet } from 'react-router-dom';
+import { getUsers } from 'utils/API/get-users';
+import './Layout.css';
+const Layout = () => {
+  const { token } = useAppSelector((state) => state.auth);
+  const appDispatch = useAppDispatch();
+  useEffect(() => {
+    token && appDispatch(getUsers());
+  }, [token, appDispatch]);
+  return (
```
Comment (ardashevdanil) [inline src/components/Layout/Layout.tsx:15]: Переносов строки не хватает, с ними было бы удобнее читать

File: src/components/ModalConfirm/ModalConfirm.tsx [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+  isError: string;
+  usersTeam: User[];
+  boards: Board[];
+  columns: Column[];
+  teammates: string[];
+  boardId: string;
+  tasks: Task[];
+};
+
+export const showDeleteConfirm = (
```
Comment (ardashevdanil) [inline src/components/ModalConfirm/ModalConfirm.tsx:69]: Странно что лежи в компонентах, хотя вроде компонента тут нет, только функция

File: src/components/ModalCreateTask/ModalCreateTask.Window.tsx [hunk 7]
```diff
@@ -0,0 +1,62 @@
+import React, { useState } from 'react';
+import { PlusOutlined } from '@ant-design/icons';
+import { Button, Modal } from 'antd';
+import { AddModalFormCreateTask } from './ModalCreateTask.Form';
+import './ModalCreateTask.scss';
+
+type PropsModal = {
+  typeButton: 'link' | 'text' | 'ghost' | 'default' | 'primary' | 'dashed' | undefined;
```
Comment (ardashevdanil) [inline src/components/ModalCreateTask/ModalCreateTask.Window.tsx:8]: Где то я это уже видел в проекте. Лучше не копипастить, а выносить в отдельные типы

File: src/pages/Board/Board.tsx [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+import { useTranslation } from 'react-i18next';
+
+export const JSONErrorHandler = (title: string, expect: string) => {
+  try {
+    const obj = JSON.parse(title);
+    return obj[expect];
+  } catch {
+    return '';
+  }
+};
```
Comment (ardashevdanil) [inline src/pages/Board/Board.tsx:33]: Можно было вынести в хелперы, вдруг где ещё пригодится

File: src/pages/Board/Board.tsx [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+      return;
+    }
+    newColumns.splice(source.index, 1);
+    newColumns.splice(destination.index, 0, columns[source.index]);
+    const sortedColumns = newColumns.map((col, idx) => ({ ...col, order: idx }));
+    dispatch(swapColumns(sortedColumns));
+    sortedColumns.map((column) => {
+      dispatch(updateBoardColumnTitle({ ...column, columnId: column._id, isSwap: true }));
+    });
+  };
```
Comment (ardashevdanil) [inline src/pages/Board/Board.tsx:120]: Очень сложно)

**Review by ardashevdanil — APPROVED** [review]
Всё ок, единственное - с форматированием что-то не то, много где без переносов написано. Плюс, скорее всего некоторую логику можно заменить готовыми библиотеками, тот же dran'n'drop

**Pull request merged**